### PR TITLE
Uq analysis

### PR DESCRIPTION
**Pull request by adewyer**
UQ sections added to mess.py, pes.py, and kb.py. Code appears to be working from test cases. If any errors are found, or bugs come up I can correct them. Normalization of the new values is underway, I'm creating a normalization class that can be called by the mess.py file.

### CODE REVIEW

File: kinbot/parameters.py [hunk 1]
```diff
@@ -194,10 +198,26 @@ def __init__(self, file=None):
...
             'mesmer_command': 'mesmer',
 
+            # UQ KEYWORDS
+            # UQ off/on (0/1)
+            'uq': 0,                   
+            # Number of mess input files generated for each structure
+            'uq_n': 5,
+            # Max number of mess calculations running at once
+            'uq_max_runs' : 5,
+            # Uncertainty in stationary point energy, +/- 0.5 kcal/mol
```
Comment (juditzador) [inline kinbot/parameters.py]: change "stationary point energy" to "stable species energy"

File: kinbot/parameters.py [hunk 2]
```diff
@@ -194,10 +198,26 @@ def __init__(self, file=None):
...
+            'stPt_uq': 0.5,
```
Comment (juditzador) [inline kinbot/parameters.py]: You could rename this then to `well_uq`, which also applies for products of course, but stPt is misleading, see above.

File: kinbot/parameters.py (hunk 1 above)
Comment (adewyer) [reply kinbot/parameters.py]: okay, good suggestion

File: kinbot/parameters.py [hunk 3]
```diff
@@ -194,10 +198,26 @@ def __init__(self, file=None):
...
+            # Uncertainty in saddle point (TS) energy, +/- 1.0 kcal/mol
+            'barrier_uq': 1.0,
+            # Uncertainty in positive frequency values, mult/div 1.2 (1.2 = 20% uncertainty)
+            'posFreq_uq': 20,
+            # Uncertainty in negative frequency values, mult/div 1.1 (1.1 = 10% uncertainty)
+            'negFreq_uq': 10,
```
Comment (juditzador) [inline kinbot/parameters.py]: I think we should ask the user to explicitly think about multiplicative factors, so I'd prefer to have the input as 1.2 which means 0.8333 to 1.2. 
Most keywords have a structure like a_b, and not aB, so can you change `posFreq` to something else? Maybe a better and consistent name would be (based on the previous things above): `freq_uq` and `imagfreq_uq`.

File: tpl/ase_gauss_ts_search.tpl.py [hunk 4]
```diff
@@ -39,11 +39,7 @@
 mol.set_calculator(calc)
 
 try:
-    db2 = connect('{working_dir}/geoms2.db')
-    db2.write(mol, name=label)
```
Comment (juditzador) [inline tpl/ase_gauss_ts_search.tpl.py]: Are these changes related to the UQ? Was it part of a test? Please check and remove if not needed.

File: tpl/ase_gauss_ts_search.tpl.py [hunk 5]
```diff
@@ -39,11 +39,7 @@
 mol.set_calculator(calc)
 
 try:
-    db2 = connect('{working_dir}/geoms2.db')
-    db2.write(mol, name=label) 
     e = mol.get_potential_energy() # use the Gaussian optimizer (task optimize)
-    db3 = connect('{working_dir}/geoms3.db')
-    db3.write(mol, name=label)
```
Comment (juditzador) [inline tpl/ase_gauss_ts_search.tpl.py]: same comment as above

File: tpl/ase_gauss_ts_search.tpl.py [hunk 6]
```diff
@@ -39,11 +39,7 @@
...
     """
     #read the geometry from the output file
```
Comment (juditzador) [inline tpl/ase_gauss_ts_search.tpl.py]: shouldn't this be deleted?

File: tpl/ase_gauss_ts_end.tpl.py [hunk 7]
```diff
@@ -33,11 +33,7 @@
 mol = Atoms(symbols = atom, positions = geom)
 mol.set_calculator(calc)
 try:
-    db2 = connect('{working_dir}/geoms2_end.db')
-    db2.write(mol, name=label)
     e = mol.get_potential_energy() # use the Gaussian optimizer
-    db3 = connect('{working_dir}/geoms3_end.db')
-    db3.write(mol, name=label)
```
Comment (juditzador) [inline tpl/ase_gauss_ts_end.tpl.py]: same comment as in the case of ase_gauss_ts_search.tpl.py

File: kinbot/zmatrix.py [hunk 8]
```diff
@@ -15,7 +15,6 @@ def make_zmat_from_cart(species, rotor, cart, mode):
     If mode = 1: only those bonds, which generate conformers
     If mode = 2: suply your one rotor in rotor as a list of atom indices
     """
-    #print(species.chemid)
```
Comment (juditzador) [inline kinbot/zmatrix.py]: delete this

File: kinbot/zmatrix.py [hunk 9]
```diff
@@ -115,8 +114,6 @@ def make_zmat_from_cart(species, rotor, cart, mode):
     for i in range(natom):
         if i == a or i == b or i == c or i == d:
             continue
-        #print(species.chemid, i, j, len(atom), len(zmat_atom))
-        #print(atom[i])
```
Comment (juditzador) [inline kinbot/zmatrix.py]: delete these

File: kinbot/parameters.py (hunk 2 above)
Comment (adewyer) [reply kinbot/parameters.py]: changed in code.

File: tpl/ase_gauss_ts_search.tpl.py (hunk 4 above)
Comment (adewyer) [reply tpl/ase_gauss_ts_search.tpl.py]: yes this should be changed. I changed in the working code on keck, but not Rosalind where the UQ code is. Its been updated.

File: tpl/ase_gauss_ts_search.tpl.py (hunk 5 above)
Comment (adewyer) [reply tpl/ase_gauss_ts_search.tpl.py]: removed.

File: tpl/ase_gauss_ts_search.tpl.py (hunk 6 above)
Comment (adewyer) [reply tpl/ase_gauss_ts_search.tpl.py]: removed.

File: tpl/ase_gauss_ts_end.tpl.py (hunk 7 above)
Comment (adewyer) [reply tpl/ase_gauss_ts_end.tpl.py]: also removed.

File: kinbot/zmatrix.py (hunk 8 above)
Comment (adewyer) [reply kinbot/zmatrix.py]: odd, but this line was not in my code. Let me know I it appears in the next push.

File: kinbot/zmatrix.py (hunk 9 above)
Comment (adewyer) [reply kinbot/zmatrix.py]: my matrix.py file has no print lines, so it is odd that they show on here. let me know if you still see them on the next push.

File: kinbot/kb.py [hunk 10]
```diff
@@ -81,6 +81,15 @@ def main():
         logging.error('Specific reaction cannot be searched in PES mode.')
         return
 
+    uq = par.par['uq']
+    uq_n=1
```
Comment (juditzador) [inline kinbot/kb.py]: spaces around the =

File: kinbot/kb.py [hunk 11]
```diff
@@ -81,6 +81,15 @@ def main():
...
+    if uq == 1:
+        uq_n = par.par['uq_n']
+        stPt_uq = par.par['stPt_uq']
```
Comment (juditzador) [inline kinbot/kb.py]: variable name needs to be changed if you changed in paramaters.py (just a note)

File: kinbot/kb.py [hunk 12]
```diff
@@ -81,6 +81,15 @@ def main():
...
+        barrier_uq = par.par['barrier_uq']
+        posFreq_uq = par.par['posFreq_uq']
+        negFreq_uq = par.par['negFreq_uq']
```
Comment (juditzador) [inline kinbot/kb.py]: same note

File: kinbot/kb.py [hunk 13]
```diff
@@ -176,18 +185,47 @@ def main():
         logging.info('Starting the search for homolytic scission products')
         well0.homolytic_scissions = HomolyticScissions(well0, par, qc)
         well0.homolytic_scissions.find_homolytic_scissions()
-  
     # initialize the master equation instance
-    mess = MESS(par, well0)
-    mess.write_input()
-    #mesmer = MESMER(par, well0)
-    #mesmer.write_input()
```
Comment (juditzador) [inline kinbot/kb.py]: How is this affecting the connection to MESMER? I think we still want to maintain that option.

File: kinbot/kb.py [hunk 14]
```diff
@@ -176,18 +185,47 @@ def main():
...
+
+    # check to see if uq analysis is turned on
+    if par.par['me'] == 1:
+        logging.info('ME turned on')
+        if uq == 0:
+            n=1
```
Comment (juditzador) [inline kinbot/kb.py]: space

File: kinbot/kb.py (hunk 13 above)
Comment (juditzador) [reply kinbot/kb.py]: OK, I misread this, these are deleted lines. Ignore this comment.

File: kinbot/kb.py [hunk 15]
```diff
@@ -176,18 +185,47 @@ def main():
...
+            logging.info('uq turned off')
+            mess = MESS(par, well0)
+            mess.write_input(uq,n)
+        elif uq == 1:
+            logging.info('uq turned on')
+            n=uq_n
```
Comment (juditzador) [inline kinbot/kb.py]: space

File: kinbot/kb.py [hunk 16]
```diff
@@ -176,18 +185,47 @@ def main():
...
+            logging.error('Cannot recognize uq code {}'.format(par.par['uq']))
+    else:
+        logging.info('ME turned off')
+    
+    #TO DO: CODE UQ into MESMER
+
+    mesmer = MESMER(par, well0)
+    mesmer.write_input()
+    #mess = MESS(par, well0)
+    #mess.write_input()
```
Comment (juditzador) [inline kinbot/kb.py]: can be deleted

File: kinbot/kb.py [hunk 17]
```diff
@@ -176,18 +185,47 @@ def main():
...
+
     if par.par['me'] == 1:
         logging.info('Starting Master Equation calculations')
         if par.par['me_code'] == 'mess':
-            mess.run()
-        elif par.par['me_code'] == '#mesmer':
+                n=uq_n
```
Comment (juditzador) [inline kinbot/kb.py]: space

File: kinbot/mess.py [hunk 18]
```diff
@@ -1,27 +1,44 @@
...
+
     def __init__(self, par, species):
         self.par = par
         self.species = species
         self.well_names = {}
         self.bimolec_names = {}
         self.fragment_names = {}
         self.ts_names = {}
-        self.termolec_names = {}
-        self.barrierless_names = {}
```
Comment (juditzador) [inline kinbot/mess.py]: You deleted this and large chunks related to termolecular and barrierless reactions. What is it about?

File: kinbot/mess.py [hunk 19]
```diff
@@ -52,71 +69,41 @@ def create_short_names(self):
...
-                                self.fragment_names[st_pt.chemid] = 'fr_' + str(len(self.fragment_names)+1)
-             #           termol_name = '_'.join(sorted([str(st_pt.chemid) for st_pt in reaction.products]))
-             #           if termol_name not in self.termolec_names:
-             #               self.termolec_names[termol_name] = 't_' + str(len(self.termolec_names)+1)
-        except:
-            logging.info("No Homolytic Scission Reactions")
-
-    def write_input(self):
+
+    def write_input(self,uq,n):
```
Comment (juditzador) [inline kinbot/mess.py]: spaces after commas

File: kinbot/mess.py [hunk 20]
```diff
@@ -52,71 +69,41 @@ def create_short_names(self):
...
         """
         write the input for all the wells, bimolecular products and barriers
         both in a separate file, as well as in one large ME file
         """
+
+        uq=uq
+        n=n
```
Comment (juditzador) [inline kinbot/mess.py]: spaces around = everywhere here
What is `uq = uq` and `n = n` about?

File: kinbot/reaction_generator.py [hunk 21]
```diff
@@ -337,11 +337,6 @@ def generate(self):
                                         prod_opt = obj.prod_opt[j]
                                         break
 
-                    elog=open("energy.log",'a')
-                    for prod_opt in obj.prod_opt:
-                        elog.write("prod_opt: {} |\tenergy: {}\n".format(prod_opt.species.chemid, prod_opt.species.energy))
-                    elog.close()
-
```
Comment (juditzador) [inline kinbot/reaction_generator.py]: Why is this deleted?

File: kinbot/kb.py (hunk 10 above)
Comment (adewyer) [reply kinbot/kb.py]: fixed.

File: kinbot/kb.py (hunk 11 above)
Comment (adewyer) [reply kinbot/kb.py]: resolved.

File: kinbot/kb.py (hunk 14 above)
Comment (adewyer) [reply kinbot/kb.py]: fixed.

File: kinbot/kb.py (hunk 15 above)
Comment (adewyer) [reply kinbot/kb.py]: fixed

File: kinbot/kb.py (hunk 16 above)
Comment (adewyer) [reply kinbot/kb.py]: I deleted a few lines here related to UQ with Mesmer that were commented out as well.

File: kinbot/kb.py (hunk 17 above)
Comment (adewyer) [reply kinbot/kb.py]: fixed

File: kinbot/parameters.py (hunk 3 above)
Comment (adewyer) [reply kinbot/parameters.py]: changed in the code

File: kinbot/mess.py (hunk 18 above)
Comment (adewyer) [reply kinbot/mess.py]: I added this information back in, I think the mess file I was working off of was from prior to the additions of barrierless and termolecular blocks to the mess.py file.

File: kinbot/reaction_generator.py (hunk 21 above)
Comment (adewyer) [reply kinbot/reaction_generator.py]: I believe that was for testing purposes to see if the right energies were being recorded. I can add it back in, but I'm not sure that the output file is necessary for people to have.
